File: vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int shift(char c);

int main(int argc, string argv[])
{
    bool error = 1;
    if (argc == 2)
    {
        error = 0;
        string keyword = argv[1];
        for (int i = 0, n = strlen(keyword); i < n; i++)
        {
            if (!isalpha(keyword[i]))
            {
                error = 1;
            }
        }
        
        if (!error)
        {
            int num_plus;
            int key;
            int cnt = 0;
            string text = get_string("plaintext: ");
            for (int i = 0, n = strlen(text); i < n; i++)
            {
                //find out what the actual key is
                key =  shift(keyword[cnt % strlen(keyword)]);
                num_plus = 0;
                if (islower(text[i]))
                {
                    num_plus = 97;
                }
                else if (isupper(text[i]))
                {
                    num_plus = 65;
                }
                if (num_plus)
                {
                    text[i] = (text[i] + key - num_plus) % 26 + num_plus;
                    cnt++;
                }
            }
            printf("ciphertext: %s\n", text);
            return 0;
        }
    }
    
    if (error == 1)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
}

int shift(char c)
{
    int num_plus = 0;
    if (islower(c))
    {
        num_plus = 97;
    }
    else if (isupper(c))
    {
        num_plus = 65;
    }
    if (num_plus)
    {
        c -= num_plus;
    }
    return c;
}

File: vigenere.py
from cs50 import get_string
from sys import argv


def char_to_int(c):
    if c.isupper():
        plus = 65
    else:
        plus = 97
    return plus


if len(argv) != 2 or not str(argv[1]).isalpha():
    exit("Usage: python vigenere.py k")
plaintext = get_string("plaintext:  ")
cypher = ""
i = 0
k = str(argv[1])
k_len = len(k)
for c in plaintext:
    if c.isalpha():
        i = i % (k_len)
        key = k[i:i+1]
        key_plus = char_to_int(key)
        key = ord(key) - key_plus
        i = i + 1
        new_c = c
        plus = char_to_int(new_c)
        new_c = ord(new_c)
        new_c = chr((new_c - plus + key) % 26 + plus)
    else:
        new_c = c
    cypher = cypher + new_c
print(f"ciphertext: {cypher}")